Human v human

lpenzey opened the pull request:
The player can now select between game modes (human vs human or human vs computer) at the start of the game. Also in this pr are a few refactors of various modules:

1.  The IO wrapper that handles `gets` and `puts` is in its own dedicated module (`TTT.IO`)
2. With the IO wrapper separated from `Input`, the rest of `Input` and `Validity` have merged to become `Validation`
3. `Output` has become `Messages` so as not to be confused with `TTT.IO`

lpenzey committed:
Relocate all input related functions to Input

lpenzey committed:
Refactor IO wrapper to dedicated module

Before the IO wrapper was mixed with modification and validation of
input. Placing it in a wrapper allows for a clearer separation of
concerns and an easier ability to redirect input/output.

lpenzey committed:
Rename Input and Output modules

Due to the IO wrapper being refactored to a dedicated module, Input has
been renamed to Validation and Output has been renamed to Messages for
clarity.The dependencies and mock dependencies are updated to reflect
this change.

lpenzey committed:
Merge HumanPlayer and ComputerPlayer

There was duplication between HumanPlayer and ComputerPlayer modules,
and at the end of the day a player simply picks a move which will return
a number, whether that move is determined by a human or by the computer.

lpenzey committed:
Add Travis CI configuration

lpenzey committed:
Add feature to choose game mode

When the game starts the user is presented with a choice to either play
against another human or against the computer player. Based on their
selection, an atom with the game mode is returned and passed into
Game.play(state, deps, mode). Game.play uses pattern matching based on
the mode to switch between the two game types.

davetorre commented on lib/TTT/game.ex:
```diff
@@ -1,38 +1,56 @@
+    move(state, deps)
+    |> computer_move()
+    |> check_status(deps, :human_v_computer)
+  end
```
Are overloaded functions an elixir convention? I think they might be but wouldn't mind having a discussion about it either way (here or in person). Seems like having two `play` functions implies that the way to add a new player configuration is to write another `play` function. Might be nicer to work with (here's where the discussion comes in) if the same `play` could handle any player configuration.

lpenzey replied:
From what I understand overloaded functions are sometimes convention. I originally had the two game modes pivoting in a case statement within one `play` function, but then separated them into two functions that used guard clauses. I do agree with you that it would be nicer to have one function that handled any player configuration, but I couldn't figure out how to do it without some form of conditional logic

esmevane replied:
Overloading in Elixir _is_ a convention, it uses pattern matching under the hood to infer the correct function.

But!  I definitely agree.  This specific case it seems like the only difference is that `computer_move()` is inserted in the pipeline.  I can second your intuition @davetorre  that it would probably be worth it to explore other approaches here if only for having that conversation.

davetorre commented on lib/TTT/game.ex:
```diff
@@ -1,38 +1,56 @@
+    deps.io.retrieve(deps.messages.get_message(:choose))
+    |> deps.validation.sanitized_move()
+    |> deps.player.analyze(state, deps)
   end
```
Might be clearer if this was renamed human_move or console_move, since there's a computer_move also

davetorre replied:
Might be clearer if this was renamed human_move or console_move, since there's a computer_move also

davetorre commented on lib/TTT/player.ex:
```diff
@@ -0,0 +1,58 @@
+    token =
+      deps.io.retrieve(deps.messages.get_message(:choose_token))
+      |> deps.validation.clean()
+      |> String.upcase()
```
Can't get enough of this pipe operator.

davetorre commented on test/game_test.exs:
```diff
@@ -32,21 +32,83 @@ defmodule GameTest do
+               },
+               player: "O"
+             }
+    end
```
Since the `computer_move` function here just calls through to the `ComputerPlayer.make_move` function, I'd probably move this test into that test file and test `make_move` directly.
Also related to this, as a general guideline I try and keep tests of a given module focused on the functions in that module that other modules depend on ("public" functions). It doesn't always work out, but sometimes if I notice myself writing tests of a private function, it indicates to me that my module might be getting too big and that the private function is showing me a potential seam where I could extract a new module.

lpenzey replied:
That's a great point - I've been meaning to ask about what the general process is for TDD'ing as it relates to private functions - is it best to just test the behavior of a public function, and not worry about the details of the private functions it relies on as long as the desired behavior is produced by the public function?

esmevane replied:
I definitely avoid testing private things.  Essentially, if the private function X is being used by Y, and changing X would change Y's output, then any test written for Y could prove that X worked, or provide some confidence, for me, about what X does.

lpenzey replied:
Makes sense, I'll start cleaning up the test files/making private functions actually private

esmevane commented on lib/TTT/game.ex:
```diff
@@ -1,38 +1,56 @@
   end
 
+  def select_mode(mode) when mode == "1", do: :human_v_human
+  def select_mode(mode) when mode == "2", do: :human_v_computer
```
👏

esmevane commented on lib/TTT/player.ex:
```diff
@@ -0,0 +1,58 @@
+         {:ok, :is_open} <- deps.validation.open?(move, state) do
+      place_move(move, state)
+    else
+      {:error, :nonexistant_space} ->
```
Oh hey, I didn't know you could pattern match on the `else` of a `with`.  Cool!

lpenzey replied:
Yeah! After working in Elixir for a little bit I want `with` along with the pipe operator in every language

esmevane commented on test/player_test.exs:
```diff
@@ -138,16 +180,42 @@ defmodule ComputerPlayerTest do
+  end
+
+  test "rejects invalid input until user selects \"O\"" do
+    Helpers.Stack.setup(["foo", "10\n", "1\n", "4\n", "o\n"])
```
Not really feedback, so forgive me for that, but I'm curious:  How do you like this `Helpers` pattern?

lpenzey replied:
I've found it to be pretty useful overall, especially for testing validation functions which make recurring calls to get input until something valid is passed in.

esmevane reviewed (approved):
This is overall great stuff!  I love a lot of your organizational tuning in this PR, and your git discipline is on point 🤘   Really well done

lpenzey committed:
Make helper functions private

lpenzey committed:
Rename move to human_move

This function was renamed for clarity so that it's clearly distinguished
from computer_move.

lpenzey committed:
Make helper functions private

rmulhol commented on config/test.exs:
```diff
@@ -1,8 +1,8 @@
-  validity: Validity,
-  human_player: MockHumanPlayer
+  validation: MockValidation,
+  messages: MockMessages,
```
It seems a little odd to me that we would need mock messages/validation. Wouldn't the mock IO take care of needed testing stubs?

rmulhol commented on lib/TTT/state.ex:
```diff
@@ -29,36 +29,14 @@ defmodule State do
   end
 
-  def switch_player("X"), do: "O"
+  defp switch_player("X"), do: "O"
```
I imagine this might be tricky to change when you support arbitrary tokens :)

rmulhol commented on config/dev.exs:
```diff
@@ -1,8 +1,8 @@
-  output: Output,
-  validity: Validity,
-  human_player: HumanPlayer
+  validation: Validation,
```
mentioned this IRL but I'm not sure if these are really deps if they live inside the project. Maybe we can get away with only configuring `:console_io`?

rmulhol reviewed (commented):
Looking good!

lpenzey committed:
Refactor IO dependency

lpenzey committed:
Extract player tokens to module attributes

In order to check for a winning combination regardless of player symbol
the player tokens have been promoted to module attributes.

lpenzey committed:
Remove project owned modules from dependencies

All modules defined by the project have been removed from dependencies
leaving only Elixir's IO module as a dependency.
